rfd/xxxx-gcp-server-discovery.md

```diff
@@ -0,0 +1,285 @@
+```
+
+The `google.compute_engine.project_id` and `google.compute_engine.zone` JWT
+claims will be mapped to configuration values.
+
+teleport.yaml on the nodes should be configured so that they will use the GCP
+join token:
+
+```yaml
+teleport:
+  join_params:
+    token_name: 'example_gcp_token'
+    method: gcp
+```
+
+### teleport.yaml generation
+
+The `teleport node configure` subcommand will be used to generate a
+new /etc/teleport.yaml file:
+
+```sh
+teleport node configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --token="$1" # name of the join token, passed via parameter from run-command
+    --labels="teleport.internal/projectId=${PROJECT_ID},\
+    teleport.internal/zone=${ZONE},\
+    teleport.internal/discovered-node=yes,\
+    teleport.internal/discovered-by=$2,\
+    teleport.internal/origin=cloud" # sourced from instance metadata
+```
+
+This will generate a file with the following contents:
+
+```yaml
+teleport:
+  nodename: '<project-id>-<zone>-<instance-id>'
+  auth_servers:
+    - 'auth-server.example.com:3025'
+  join_params:
+    token_name: token
+  # ...
+ssh_service:
+  enabled: 'yes'
+  labels:
+    teleport.internal/projectId: '<project-id>'
+    teleport.internal/zone: '<zone>'
+```
+
+## UX
+
+### User has 1 account to discover servers on
+
+#### Teleport config
+
+Discovery server:
+
+```yaml
+teleport: ...
+auth_service:
+  enabled: 'yes'
+discovery_service:
+  enabled: 'yes'
+  gcp:
+    - types: ['vm']
+      project_ids: ['<project_id>']
+      locations: ['westcentralus']
+      tags:
+        'teleport': 'yes'
+      install:
+        # Use default values
+```
+
+Permissions required for the Discovery Service:
+
+```yaml
+title: teleport_vm_discovery
+description: Role for discovering VMs and adding them to a Teleport cluster.
+stage: ALPHA
+includedPermissions:
+  - compute.instances.setMetadata
+  - compute.instances.get
+  - compute.instances.list
+```
+
+Teleport will retrieve GCP credentials in the same way that it already does for
+[GKE auto-discovery](https://goteleport.com/docs/kubernetes-access/discovery/google-cloud/#retrieve-credentials-for-your-teleport-services).
+
+## Security considerations
+
+Automatic EC2 joining uses SSM to separate permission to create/update commands
+from permission to call them. GCP does not have an SSM-like service, so the
+discovery service will require permission to both create and execute commands.
+
+GCP does not have a dedicated "Run command" API; commands on VMs are instead
+executed over SSH, with access managed with keys in the VM's
+metadata. There is no permission specifically for running commands; the most
+limiting one is `compute.instances.setMetadata`.
+
+## Links
+
+- [Verifying GCP VM Identity](https://cloud.google.com/compute/docs/instances/verifying-instance-identity)
+- [SSH on GCP VMs](https://cloud.google.com/compute/docs/instances/ssh#third-party-tools)
+
+## Appendix I - Example ID token payload
```

Comment:
Can you also include IaC flows in the scope of the RFD (whether any Helm charts, Kube operator or Terraform provider changes will be needed).